### PR TITLE
Implement consul datasource

### LINKED ISSUE #75: Add consul as dynamic datasource
## Issue Description

Type: *feature request*

### PR DESCRIPTION
**Pull request by gorexlv**
### Describe what this PR does / why we need it

implement consul datasource

### Describe how you did it

adapter consul client api;
read the configuration immediately after connecting to consul server,
then start polling.

### Special notes for reviews

### CODE REVIEW

**Comment by gorexlv** [comment]
please pause the merge, more detail testing is in progress.

Is there any way to switch to draft PR?

**Comment by louyuting** [comment]
> please pause the merge, more detail testing is in progress.
> 
> Is there any way to switch to draft PR?

Can not.
You could create Draft PR. Once you creates OPNE PR, you can not switch to Draft.

**Comment by louyuting** [comment]
By the way, The CI failed, could you please fix it?

Thanks.

File: ext/datasource/consul/consul.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+	for _, h := range options.propertyHandlers {
+		ds.AddPropertyHandler(h)
+	}
+	return ds
+}
+
+// ReadSource implement datasource.DataSource interface
+func (c *consulDataSource) ReadSource() ([]byte, error) {
+	pair, meta, err := c.kvQuerier.Get(c.propertyKey, c.queryOptions)
```
Comment (louyuting) [inline ext/datasource/consul/consul.go:67]: We might need to check nil for pair and meta.

File: ext/datasource/consul/consul.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+	if err != nil {
+		logger.Errorf("[consul] ReadSource failed: %s", err.Error())
+		return nil, err
+	}
+
+	c.waitIndex = meta.LastIndex
+	return pair.Value, nil
```
Comment (louyuting) [inline ext/datasource/consul/consul.go:75]: It mights cause nil pointer if key is not existed.

File: ext/datasource/consul/consul.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+	logger = logging.GetDefaultLogger()
+)
+
+// NewConsulDataSource returns new consul datasource instance
+func NewConsulDataSource(propertyKey string, opts ...Option) (datasource.DataSource, error) {
+	var options = evaluateOptions(opts)
+	// if not consul client is specified, initialize from the configuration
+	if options.consulClient == nil {
+		client, err := api.NewClient(options.consulConfig)
```
Comment (louyuting) [inline ext/datasource/consul/consul.go:37]: It mights cause nil pointer if options.consulConfig don't be assigned.
Besides, Is it better to force the user to pass in the client parameter manually?

File: ext/datasource/consul/consul.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	"github.com/alibaba/sentinel-golang/ext/datasource"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/cenkalti/backoff"
+	"github.com/hashicorp/consul/api"
+)
+
+type consulDataSource struct {
+	datasource.Base
+	*options
```
Comment (louyuting) [inline ext/datasource/consul/consul.go:16]: Maybe this field(*options) is unnecessary？

File: ext/datasource/consul/consul_example_test.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+import (
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/hashicorp/consul/api"
+)
+
+func Example_ConsulDataSource_CustomizeClient() {
+	client, err := api.NewClient(nil)
```
Comment (louyuting) [inline ext/datasource/consul/consul_example_test.go:11]: It causes nil pointer.

File: ext/datasource/consul/option.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+// build a new backoff instance for every consul query
+func WithBackOffBuilder(backoffBuilder func() backoff.BackOff) Option {
+	return func(opts *options) {
+		opts.backoffBuilder = backoffBuilder
+	}
+}
+
+func evaluateOptions(opts []Option) *options {
+	var optCopy = &options{
+		propertyHandlers: make([]datasource.PropertyHandler, 0),
```
Comment (louyuting) [inline ext/datasource/consul/option.go:81]: Watcher on by default is even better

File: ext/datasource/consul/consul.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	if !c.disableWatch {
+		go util.RunWithRecover(c.watch, logger)
+	}
+	return nil
+}
+
+func (c consulDataSource) watch() {
+	var waitIndex uint64
```
Comment (louyuting) [inline ext/datasource/consul/consul.go:94]: maybe init waitIndex as c. waitIndex?

File: ext/datasource/consul/consul.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		logger.Errorf("[consul] doReadAndUpdate failed: %s", err.Error())
+	}
+
+	if !c.disableWatch {
+		go util.RunWithRecover(c.watch, logger)
+	}
+	return nil
+}
+
+func (c consulDataSource) watch() {
```
Comment (louyuting) [inline ext/datasource/consul/consul.go:93]: Maybe pointer receiver?

**Comment by louyuting** [comment]
Could you please rebase the code? 
CI shows the conflicts.
Thanks

File: ext/datasource/consul/consul.go (hunk 1 above)
Comment (gorexlv) [reply ext/datasource/consul/consul.go]: query must be checked but meta is safe.

**Comment by gorexlv** [comment]
#159

File: ext/datasource/consul/consul.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+func (c *consulDataSource) doReadAndUpdate() (err error) {
+	src, err := c.ReadSource()
+	if err != nil {
+		return err
+	}
+	return c.doUpdate(src)
+}
+
+func (c *consulDataSource) Close() error {
+	if c.cancel != nil {
```
Comment (louyuting) [inline ext/datasource/consul/consul.go:135]: I run the test:
```
func Test_consulDatasource_CustomizeClient(t *testing.T) {
	client, err := api.NewClient(&api.Config{
		Address: "127.0.0.1:8500",
	})
	if err != nil {
		// todo something
	}
	ds, err := NewDatasource("property_key",
		// customize consul client
		WithConsulClient(client),
		// disable dynamic datasource watch
		WithDisableWatch(false),
		// preset property handlers
		WithPropertyHandlers(),
		// reset queryOptions, defaultQueryOptions as default
		WithQueryOptions(&api.QueryOptions{}),
	)

	if err != nil {
		// todo something
	}

	if err := ds.Initialize(); err != nil {
		// todo something
	}
	time.Sleep(time.Second * 10)

	ds.Close()

	time.Sleep(time.Second * 1000)
}
```

But the cancel don't work as expected.

Is there anything wrong in my test?

Comment (gorexlv) [reply ext/datasource/consul/consul.go:135]: Maybe the lastest code line `time.Sleep(time.Second*1000)` causes you to think that there is no cancel.

I'll improve the unit test case later.

File: ext/datasource/consul/consul.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+	"github.com/alibaba/sentinel-golang/ext/datasource"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/hashicorp/consul/api"
+)
+
+type consulDataSource struct {
+	datasource.Base
+	*options
```
Comment (louyuting) [inline ext/datasource/consul/consul.go:16]: It seems this field is useless for datasource, 
Deleting this field makes more sense.

File: ext/datasource/consul/consul_example_test.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+		Address: "127.0.0.1:8500",
+	})
+	if err != nil {
+		// todo something
+	}
+	ds, err := NewDatasource("property_key",
+		// customize consul client
+		WithConsulClient(client),
+		// disable dynamic datasource watch
+		WithDisableWatch(true),
```
Comment (sczyh30) [inline ext/datasource/consul/consul_example_test.go:20]: Maybe setting false is better here?

**Comment by sczyh30** [comment]
@louyuting We may also need to add out-of-box demo for various data-sources in `example` pkg. Just `xxx_example_test` is not enough.

**Comment by louyuting** [comment]
> @louyuting We may also need to add out-of-box demo for various data-sources in `example` pkg. Just `xxx_example_test` is not enough.

I think so too. We might need to add those demo into `example` pkg.

**Pull request merged**